// @mastercard/ng-commons/bundles/mastercard-ng-commons.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@ngx-translate/core'), require('@angular/http'), require('rxjs/operators'), require('rxjs/observable/forkJoin'), require('@angular/router'), require('rxjs/observable/of'), require('rxjs/observable/fromEvent'), require('moment'), require('@angular/common')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/forms', '@ngx-translate/core', '@angular/http', 'rxjs/operators', 'rxjs/observable/forkJoin', '@angular/router', 'rxjs/observable/of', 'rxjs/observable/fromEvent', 'moment', '@angular/common'], factory) :
	(factory((global.mastercard = global.mastercard || {}, global.mastercard['ng-commons'] = {}),global.ng.core,global.ng.forms,global['@ngx-translate/core'],global.ng.http,global.Rx.Observable.prototype,global.Rx.Observable,global.ng.router,global.Rx.Observable,global.Rx.Observable,global.moment,global.ng.common));
}(this, (function (exports,core,forms,core$1,http,operators,forkJoin,router,of,fromEvent,momentNS,common) { 'use strict';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0
THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.
See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */










function __read(o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
}
function __spread() {
    for (var ar = [], i = 0; i < arguments.length; i++)
        ar = ar.concat(__read(arguments[i]));
    return ar;
}

var CustomValidatorDirective = /** @class */ (function () {
    function CustomValidatorDirective() {
    }
    CustomValidatorDirective.prototype.validate = function (c) {
        return this.mcCustomValidator(c);
    };
    return CustomValidatorDirective;
}());
CustomValidatorDirective.decorators = [
    { type: core.Directive, args: [{
                selector: '[mcCustomValidator][ngModel]',
                providers: [
                    { provide: forms.NG_VALIDATORS, useExisting: core.forwardRef(function () { return CustomValidatorDirective; }), multi: true }
                ]
            },] },
];
CustomValidatorDirective.ctorParameters = function () { return []; };
CustomValidatorDirective.propDecorators = {
    "mcCustomValidator": [{ type: core.Input },],
};
var ShowErrorIconDirective = /** @class */ (function () {
    function ShowErrorIconDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.hasError = false;
        this.errorIconElement = this.renderer.createElement("span");
        this.divElement = this.renderer.createElement("div");
    }
    ShowErrorIconDirective.prototype.ngOnInit = function () {
        var parent = this.el.nativeElement.parentNode;
        this.renderer.addClass(this.errorIconElement, "sprite");
        this.renderer.addClass(this.errorIconElement, "icon-error");
        this.renderer.addClass(this.divElement, "input-icon-error-wrapper");
        this.renderer.insertBefore(parent, this.divElement, this.el.nativeElement);
        this.renderer.removeChild(parent, this.el.nativeElement);
        this.renderer.appendChild(this.divElement, this.el.nativeElement);
    };
    ShowErrorIconDirective.prototype.ngOnChanges = function () {
        if (this.mcShowErrorIcon) {
            this.addError();
        }
        else {
            this.removeError();
        }
    };
    ShowErrorIconDirective.prototype.addError = function () {
        this.renderer.appendChild(this.divElement, this.errorIconElement);
        this.hasError = true;
    };
    ShowErrorIconDirective.prototype.removeError = function () {
        if (this.hasError) {
            this.renderer.removeChild(this.divElement, this.errorIconElement);
            this.hasError = false;
        }
    };
    return ShowErrorIconDirective;
}());
ShowErrorIconDirective.decorators = [
    { type: core.Directive, args: [{
                selector: '[mcShowErrorIcon]'
            },] },
];
ShowErrorIconDirective.ctorParameters = function () { return [
    { type: core.ElementRef, },
    { type: core.Renderer2, },
]; };
ShowErrorIconDirective.propDecorators = {
    "mcShowErrorIcon": [{ type: core.Input },],
};
function formFactory(form) {
    return form;
}
var ProvideParentFormDirective = /** @class */ (function () {
    function ProvideParentFormDirective() {
    }
    return ProvideParentFormDirective;
}());
ProvideParentFormDirective.decorators = [
    { type: core.Directive, args: [{
                selector: '[mcProvideParentForm]',
                providers: [
                    {
                        provide: forms.ControlContainer,
                        useFactory: formFactory,
                        deps: [forms.NgForm]
                    }
                ]
            },] },
];
ProvideParentFormDirective.ctorParameters = function () { return []; };
var LogLevel = {
    DEBUG: 0,
    INFO: 1,
    WARN: 2,
    ERROR: 3,
};
LogLevel[LogLevel.DEBUG] = "DEBUG";
LogLevel[LogLevel.INFO] = "INFO";
LogLevel[LogLevel.WARN] = "WARN";
LogLevel[LogLevel.ERROR] = "ERROR";
var logLevel = LogLevel.INFO;
function getDefaultLogLevel() {
    return logLevel;
}
function setDefaultLogLevel(level) {
    logLevel = level;
}
var Logger = /** @class */ (function () {
    function Logger(level) {
        this.level = level;
    }
    Logger.prototype.error = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        if (this.level <= LogLevel.ERROR)
            console.error.apply(console, __spread([message], optionalParams));
    };
    Logger.prototype.warn = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        if (this.level <= LogLevel.WARN)
            console.warn.apply(console, __spread([message], optionalParams));
    };
    Logger.prototype.info = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        if (this.level <= LogLevel.INFO)
            console.info.apply(console, __spread([message], optionalParams));
    };
    Logger.prototype.debug = function (message) {
        var optionalParams = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParams[_i - 1] = arguments[_i];
        }
        if (this.level <= LogLevel.DEBUG)
            console.log.apply(console, __spread([message], optionalParams));
    };
    return Logger;
}());
Logger.decorators = [
    { type: core.Injectable },
];
Logger.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: core.Inject, args: ['LogLevel',] },] },
]; };
var StrictPatternDirective = /** @class */ (function () {
    function StrictPatternDirective(el, logger) {
        this.el = el;
        this.logger = logger;
        this.ngModelChange = new core.EventEmitter();
        this.logger.debug('Strict Pattern Directive built');
    }
    StrictPatternDirective.prototype.onChange = function (e) {
        if (this.event) {
            if (this.event === 'change') {
                this.validation(e);
            }
            else {
                this.logger.warn('Only change event is allowed.');
            }
        }
    };
    StrictPatternDirective.prototype.onInput = function (e) {
        if (!this.event) {
            this.validation(e);
        }
    };
    StrictPatternDirective.prototype.validation = function (e) {
        var newValue = e.target.value;
        var pattern = this.el.nativeElement.attributes.getNamedItem('pattern').value;
        var reg = new RegExp(pattern);
        if (reg.test(newValue) || newValue === '') {
            e.target.previousValue = e.target.value;
            this.logger.debug("Pattern ok. Value = \"" + e.target.value + "\"");
        }
        else {
            e.target.value = e.target.previousValue || '';
            this.ngModelChange.emit(e.target.value);
            this.logger.debug("Pattern ko. Restoring value = \"" + e.target.value + "\"");
        }
    };
    return StrictPatternDirective;
}());
StrictPatternDirective.decorators = [
    { type: core.Directive, args: [{
                selector: 'input[type=text][mcStrict][pattern]'
            },] },
];
StrictPatternDirective.ctorParameters = function () { return [
    { type: core.ElementRef, },
    { type: Logger, },
]; };
StrictPatternDirective.propDecorators = {
    "event": [{ type: core.Input, args: ['mcStrict',] },],
    "ngModelChange": [{ type: core.Output },],
    "onChange": [{ type: core.HostListener, args: ['change', ['$event'],] },],
    "onInput": [{ type: core.HostListener, args: ['input', ['$event'],] },],
};
var TrimDirective = /** @class */ (function () {
    function TrimDirective(element) {
        this.ngModelChange = new core.EventEmitter();
        this.element = element.nativeElement;
    }
    TrimDirective.prototype.onChange = function (value) {
        value = value.trim();
        this.element.value = value;
        this.ngModelChange.emit(value);
    };
    return TrimDirective;
}());
TrimDirective.decorators = [
    { type: core.Directive, args: [{
                selector: 'input[mcTrim],textarea[mcTrim]'
            },] },
];
TrimDirective.ctorParameters = function () { return [
    { type: core.ElementRef, },
]; };
TrimDirective.propDecorators = {
    "ngModelChange": [{ type: core.Output },],
    "onChange": [{ type: core.HostListener, args: ['change', ['$event.target.value'],] },],
};
function trackByIndex(i) {
    return i;
}
function searchValidator() {
    return function (c) {
        var pattern = /[\?\'\"\<\>\/]+/g;
        var value = String(c.value || '');
        return pattern.test(value) ? { invalidCharError: true } : null;
    };
}
function makeConfig(factoryFn) {
    return {
        loader: {
            provide: core$1.TranslateLoader,
            useFactory: factoryFn,
            deps: [http.Http],
        },
    };
}
var TranslateStaticLoader = /** @class */ (function () {
    function TranslateStaticLoader(http$$1, prefix, suffix, apps) {
        if (prefix === void 0) { prefix = 'i18n'; }
        if (suffix === void 0) { suffix = '.json'; }
        this.http = http$$1;
        this.prefix = prefix;
        this.suffix = suffix;
        this.apps = apps;
    }
    TranslateStaticLoader.prototype.getTranslation = function (lang) {
        var _this = this;
        var requests = [];
        var prefix = this.prefix;
        var suffix = this.suffix;
        requests.push(this.http.get(prefix + '/' + lang + '/' + lang + suffix));
        this.apps.forEach(function (appName) {
            var url = prefix + '/' + lang + '/' + appName + '_' + lang + suffix;
            requests.push(_this.http.get(url));
        });
        return forkJoin.forkJoin.apply(void 0, __spread(requests)).pipe(operators.map(function (responses) {
            var first = responses.shift();
            var combined = first && first.json();
            var counter = 0;
            responses.forEach(function (res) {
                var translates = res.json();
                var appName = _this.apps[counter];
                combined[appName.toUpperCase()] = translates[appName.toUpperCase()];
                counter++;
            });
            return combined;
        }));
    };
    return TranslateStaticLoader;
}());
var LOCALE = Object.freeze({
    ENGLISH: 'en-US',
    SPANISH: 'es-CO',
    PORTUGUESE: 'pt-BR',
    get DEFAULT() { return this.ENGLISH; },
});
var LOCALE_STORAGE_KEY = 'locale';
var LOCALE_QUERY_PARAM = 'locale';
var LocaleService = /** @class */ (function () {
    function LocaleService(window, activated) {
        this.activated = activated;
        this.window = window;
    }
    LocaleService.prototype.getCurrent = function () {
        var _this = this;
        return of.of(LOCALE.DEFAULT).pipe(operators.map(function (defaultLocale) {
            var sessionLocale = _this.window.sessionStorage.getItem(LOCALE_STORAGE_KEY);
            return sessionLocale || defaultLocale;
        }), operators.mergeMap(function (locale) {
            if (_this.activated) {
                return _this.activated.queryParams.pipe(operators.map(function (params) {
                    return params[LOCALE_QUERY_PARAM] || locale;
                }));
            }
            else {
                return of.of(locale);
            }
        }));
    };
    return LocaleService;
}());
LocaleService.decorators = [
    { type: core.Injectable },
];
LocaleService.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: core.Inject, args: ['Window',] },] },
    { type: router.ActivatedRoute, decorators: [{ type: core.Optional },] },
]; };
function filterSameOrigin(window) {
    return function (msg) { return msg.origin === window.location.origin; };
}
function filterSameWindow(source) {
    return function (msg) { return msg.source === source; };
}
var MessageService = /** @class */ (function () {
    function MessageService(window) {
        this.window = window;
        this.message$ = fromEvent.fromEvent(this.window, 'message').pipe(operators.filter(filterSameOrigin(window)));
    }
    MessageService.prototype.get = function (source) {
        if (source === void 0) { source = this.window.parent; }
        var fn = filterSameWindow(source);
        return this.message$.pipe(operators.filter(fn));
    };
    MessageService.prototype.post = function (message, target) {
        if (target === void 0) { target = this.window.parent; }
        var sameOrigin = this.window.location.origin;
        target.postMessage(message, sameOrigin);
    };
    return MessageService;
}());
MessageService.decorators = [
    { type: core.Injectable },
];
MessageService.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: core.Inject, args: ['Window',] },] },
]; };
var moment = momentNS;
var DateFormatPipe = /** @class */ (function () {
    function DateFormatPipe() {
    }
    DateFormatPipe.prototype.transform = function (value) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var _a = __read(args, 1), format = _a[0];
        if (!format) {
            format = 'DD-MMM-YYYY hh:mm A';
        }
        return moment(value).format(format);
    };
    return DateFormatPipe;
}());
DateFormatPipe.decorators = [
    { type: core.Pipe, args: [{
                name: 'dateFormat'
            },] },
];
DateFormatPipe.ctorParameters = function () { return []; };
var RequiredFieldLabelPipe = /** @class */ (function () {
    function RequiredFieldLabelPipe() {
        this.mark = "<span class='required'>*</span>";
    }
    RequiredFieldLabelPipe.prototype.transform = function (value, field) {
        var visible = typeof field === 'undefined' || field.required;
        return value && visible ? value + this.mark : value;
    };
    return RequiredFieldLabelPipe;
}());
RequiredFieldLabelPipe.decorators = [
    { type: core.Pipe, args: [{ name: 'requiredFieldLabel' },] },
];
RequiredFieldLabelPipe.ctorParameters = function () { return []; };
function getWindow() { return window; }
var declarations = [
    CustomValidatorDirective,
    TrimDirective,
    RequiredFieldLabelPipe,
    ProvideParentFormDirective,
    StrictPatternDirective,
    ShowErrorIconDirective,
    DateFormatPipe,
];
var MastercardCommons = /** @class */ (function () {
    function MastercardCommons() {
    }
    MastercardCommons.forRoot = function () {
        return {
            ngModule: MastercardCommons,
            providers: [
                Logger,
                MessageService,
                LocaleService,
                { provide: 'LogLevel', useFactory: getDefaultLogLevel },
                { provide: 'Window', useFactory: getWindow },
            ],
        };
    };
    return MastercardCommons;
}());
MastercardCommons.decorators = [
    { type: core.NgModule, args: [{
                declarations: declarations,
                imports: [common.CommonModule],
                exports: __spread(declarations),
            },] },
];
MastercardCommons.ctorParameters = function () { return []; };

exports.CustomValidatorDirective = CustomValidatorDirective;
exports.ShowErrorIconDirective = ShowErrorIconDirective;
exports.formFactory = formFactory;
exports.ProvideParentFormDirective = ProvideParentFormDirective;
exports.StrictPatternDirective = StrictPatternDirective;
exports.TrimDirective = TrimDirective;
exports.trackByIndex = trackByIndex;
exports.searchValidator = searchValidator;
exports.makeConfig = makeConfig;
exports.TranslateStaticLoader = TranslateStaticLoader;
exports.LocaleService = LocaleService;
exports.LOCALE = LOCALE;
exports.LOCALE_STORAGE_KEY = LOCALE_STORAGE_KEY;
exports.LOCALE_QUERY_PARAM = LOCALE_QUERY_PARAM;
exports.LogLevel = LogLevel;
exports.getDefaultLogLevel = getDefaultLogLevel;
exports.setDefaultLogLevel = setDefaultLogLevel;
exports.Logger = Logger;
exports.filterSameOrigin = filterSameOrigin;
exports.filterSameWindow = filterSameWindow;
exports.MessageService = MessageService;
exports.getWindow = getWindow;
exports.MastercardCommons = MastercardCommons;
exports.ɵb = DateFormatPipe;
exports.ɵa = RequiredFieldLabelPipe;

Object.defineProperty(exports, '__esModule', { value: true });

})));
