// @mc-fraud-center/stand-in-data-model/bundles/mc-fraud-center-stand-in-data-model.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('cerialize'), require('moment')) :
	typeof define === 'function' && define.amd ? define(['exports', 'cerialize', 'moment'], factory) :
	(factory((global['mc-fraud-center'] = global['mc-fraud-center'] || {}, global['mc-fraud-center']['stand-in-data-model'] = {}),global.cerialize,global.moment));
}(this, (function (exports,cerialize,momentNS) { 'use strict';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0
THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.
See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */
var extendStatics = Object.setPrototypeOf ||
    ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
    function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}


function __decorate(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}

function __metadata(metadataKey, metadataValue) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
}

var ErrorCodes = Object.freeze({
    ACCT_RANGE_NOT_ALLOWED: 'acct.range.not.allowed',
    BILLING_LIMIT_EXCEED: 'billing.limit.exceed',
    BLOCK_ACCT_RANGE_TOO_WIDE: 'block.acct.range.too.wide',
    BLOCK_DUPLICATES_ICA_BLOCK: 'block.duplicates.ica.block',
    BLOCK_DUPLICATES_RANGE_BLOCK: 'block.duplicates.range.block',
    BLOCK_NOT_FOR_CUSTOMER: 'block.not.for.cust',
    BLOCK_SHADOWS_ICA_BLOCK: 'block.shadows.ica.block',
    BLOCK_SHADOWS_RANGE_BLOCK: 'block.shadows.range.block',
    ICA_NOT_ALLOWED: 'ica.not.allowed',
    INVALID_CUSTOMER: 'invalid.customer',
    INVALID_DATA_NOT_IN_DOMAIN: 'invalid.data.not.in.domain',
    INVALID_DATA_ORDER: 'invalid.data.order',
    INVALID_DATA_SIZE: 'invalid.data.size',
    INVALID_DATA_TYPE: 'invalid.data.type',
    INVALID_DATA_VALUE_FUTURE: 'invalid.date.value.future',
    INVALID_DATA_VALUE_PAST: 'invalid.date.value.past',
    INVALID_DECIMAL_VALUE_MAX: 'invalid.decimal.value.max',
    INVALID_DECIMAL_VALUE_MIN: 'invalid.decimal.value.min',
    INVALID_MAP_DUPLICATE_KEY: 'invalid.map.duplicate.key',
    INVALID_NOT_NULL_VALUE: 'invalid.not.null.value',
    INVALID_NUMERIC_VALUE_CURRENCY: 'invalid.numeric.value.currency',
    INVALID_NUMERIC_VALUE_DIGITS: 'invalid.numeric.value.digits',
    INVALID_NUMERIC_VALUE_MAX: 'invalid.numeric.value.max',
    INVALID_NUMERIC_VALUE_MIN: 'invalid.numeric.value.min',
    INVALID_NUMERIC_VALUE_RANGE: 'invalid.numeric.value.range',
    INVALID_TEXT_VALUE_BLANK: 'invalid.text.value.blank',
    INVALID_TEXT_VALUE_EMPTY: 'invalid.text.value.empty',
    INVALID_TEXT_VALUE_LENGTH: 'invalid.text.value.length',
    INVALID_TEXT_VALUE_PATTERN: 'invalid.text.value.pattern',
    REQUEST_INVALID_ILLEGAL_ARGUMENT: 'request.invalid.illegal.argument',
    REQUEST_INVALID_MESSAGE_NOT_READABLE: 'request.invalid.message.not.readable',
    REQUEST_INVALID_UNSATISFIED_REQUEST_PARAMETER: 'request.invalid.unsatisfied.request.parameter',
    TCC_LIMIT_ACCT_RANGE_NOT_FOUND: 'tcc.limit.acct.range.not.found',
    TCC_LIMIT_MINIMUM_THRESHOLD_VIOLATION: 'tcc.limit.min.threshold.violation',
    ACCUM_LIMIT_ACCT_RANGE_NOT_FOUND: 'accum.limit.acct.range.not.found',
    ACCUM_LIMIT_MIN_AMOUNT_THRESHOLD_VIOLATION: 'accum.limit.min.amount.threshold.violation',
    ACCUM_LIMIT_MAX_AMOUNT_THRESHOLD_VIOLATION: 'accum.limit.max.amount.threshold.violation',
    ACCUM_LIMIT_MIN_COUNT_THRESHOLD_VIOLATION: 'accum.limit.min.count.threshold.violation',
    ACCUM_LIMIT_MAX_COUNT_THRESHOLD_VIOLATION: 'accum.limit.max.count.threshold.violation',
    REQUEST_UNAUTHORIZED: 'request.unauthorized',
    ACCESS_DENIED_CSRF: 'access.denied.csrf',
    ACCESS_DENIED_PERMISSION: 'access.denied.permission',
    RESOURCE_NOT_FOUND: 'resource.not.found',
    REQUEST_INVALID_METHOD: 'request.invalid.method',
    REQUEST_INVALID_MEDIA_TYPE: 'request.invalid.media.type',
    REQUEST_CONFLICT: 'request.conflict',
    RESOURCE_GONE: 'resource.gone',
    REQUEST_UNSUPPORTED_MEDIA_TYPE: 'request.unsupported.media.type',
    SYSTEM: 'system',
    SERVICE_UNAVAILABLE: 'service.unavailable'
});
var BlockCriteria = /** @class */ (function () {
    function BlockCriteria(initial) {
        Object.assign(this, initial);
    }
    return BlockCriteria;
}());
__decorate([
    cerialize.deserialize,
    __metadata("design:type", Number)
], BlockCriteria.prototype, "criteriaId", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Number)
], BlockCriteria.prototype, "elementId", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], BlockCriteria.prototype, "value", void 0);
var BlockType = {
    RANGE: 0,
    ICA: 1,
};
BlockType[BlockType.RANGE] = "RANGE";
BlockType[BlockType.ICA] = "ICA";
cerialize.SerializableEnumeration(BlockType);
var StandInBlock = /** @class */ (function () {
    function StandInBlock(initial) {
        this.blockCriteria = [];
        Object.assign(this, initial);
    }
    StandInBlock.prototype.getBlockCriteria = function (id) {
        var found = this.blockCriteria && this.blockCriteria.find(function (item) { return item.elementId === id; });
        if (found) {
            return found;
        }
        var blockCriteria = new BlockCriteria();
        blockCriteria.elementId = id;
        blockCriteria.value = null;
        return blockCriteria;
    };
    return StandInBlock;
}());
__decorate([
    cerialize.deserialize,
    __metadata("design:type", Number)
], StandInBlock.prototype, "blockId", void 0);
__decorate([
    cerialize.deserializeAs(BlockType),
    __metadata("design:type", Number)
], StandInBlock.prototype, "blockType", void 0);
__decorate([
    cerialize.deserialize,
    __metadata("design:type", Number)
], StandInBlock.prototype, "memberId", void 0);
__decorate([
    cerialize.autoserializeAs(BlockCriteria),
    __metadata("design:type", Array)
], StandInBlock.prototype, "blockCriteria", void 0);
__decorate([
    cerialize.deserializeAs(Date),
    __metadata("design:type", Date)
], StandInBlock.prototype, "lastUpdateDttm", void 0);
__decorate([
    cerialize.deserialize,
    __metadata("design:type", String)
], StandInBlock.prototype, "lastUpdateUsr", void 0);
__decorate([
    cerialize.deserializeAs(Date),
    __metadata("design:type", Date)
], StandInBlock.prototype, "crteDttm", void 0);
__decorate([
    cerialize.deserialize,
    __metadata("design:type", String)
], StandInBlock.prototype, "crteUsr", void 0);
exports.RangeBlock = /** @class */ (function (_super) {
    __extends(RangeBlock, _super);
    function RangeBlock(initial) {
        var _this = _super.call(this, initial) || this;
        Object.assign(_this, initial);
        _this.blockType = BlockType.RANGE;
        return _this;
    }
    return RangeBlock;
}(StandInBlock));
__decorate([
    cerialize.deserialize,
    __metadata("design:type", String)
], exports.RangeBlock.prototype, "startingBlockRange", void 0);
__decorate([
    cerialize.deserialize,
    __metadata("design:type", String)
], exports.RangeBlock.prototype, "endingBlockRange", void 0);
exports.RangeBlock = __decorate([
    cerialize.inheritSerialization(StandInBlock),
    __metadata("design:paramtypes", [exports.RangeBlock])
], exports.RangeBlock);
exports.IcaBlock = /** @class */ (function (_super) {
    __extends(IcaBlock, _super);
    function IcaBlock(initial) {
        var _this = _super.call(this, initial) || this;
        Object.assign(_this, initial);
        _this.blockType = BlockType.ICA;
        return _this;
    }
    return IcaBlock;
}(StandInBlock));
exports.IcaBlock = __decorate([
    cerialize.inheritSerialization(StandInBlock),
    __metadata("design:paramtypes", [exports.IcaBlock])
], exports.IcaBlock);
var moment = momentNS;
var Filters = /** @class */ (function () {
    function Filters(initial) {
        Object.assign(this, initial);
    }
    Object.defineProperty(Filters.prototype, 'start-range', {
        get: function () { return this.startRange; },
        set: function (v) { this.startRange = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'end-range', {
        get: function () { return this.endRange; },
        set: function (v) { this.endRange = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'block-type', {
        get: function () { return this.blockType; },
        set: function (v) { this.blockType = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'member-id', {
        get: function () { return this.memberId; },
        set: function (v) { this.memberId = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'pos-ctry', {
        get: function () { return this.posCtry; },
        set: function (v) { this.posCtry = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'create-dttm-start', {
        get: function () { return this.createDttmStart; },
        set: function (v) { this.createDttmStart = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'create-dttm-end', {
        get: function () {
            if (this['create-dttm-start']) {
                this.createDttmEnd = moment(this['create-dttm-start']).add(1, 'days');
            }
            else {
                this.createDttmEnd = this['create-dttm-start'];
            }
            return this.createDttmEnd;
        },
        set: function (_) {
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'tx-type', {
        get: function () { return this.txType; },
        set: function (v) { this.txType = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'pan-entry', {
        get: function () { return this.panEntry; },
        set: function (v) { this.panEntry = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'cat-level', {
        get: function () { return this.catLevel; },
        set: function (v) { this.catLevel = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'acq-id', {
        get: function () { return this.acqId; },
        set: function (v) { this.acqId = v; },
        enumerable: true,
        configurable: true
    });
    return Filters;
}());
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], Filters.prototype, "mcc", void 0);
__decorate([
    cerialize.autoserializeAs('start-range'),
    __metadata("design:type", String)
], Filters.prototype, "startRange", void 0);
__decorate([
    cerialize.autoserializeAs('end-range'),
    __metadata("design:type", String)
], Filters.prototype, "endRange", void 0);
__decorate([
    cerialize.autoserializeAs('block-type'),
    __metadata("design:type", String)
], Filters.prototype, "blockType", void 0);
__decorate([
    cerialize.autoserializeAs('member-id'),
    __metadata("design:type", String)
], Filters.prototype, "memberId", void 0);
__decorate([
    cerialize.autoserializeAs('pos-ctry'),
    __metadata("design:type", String)
], Filters.prototype, "posCtry", void 0);
__decorate([
    cerialize.autoserializeAs(Date, 'create-dttm-start'),
    __metadata("design:type", Date)
], Filters.prototype, "createDttmStart", void 0);
__decorate([
    cerialize.autoserializeAs(Date, 'create-dttm-end'),
    __metadata("design:type", Object)
], Filters.prototype, "createDttmEnd", void 0);
__decorate([
    cerialize.autoserializeAs('tx-type'),
    __metadata("design:type", String)
], Filters.prototype, "txType", void 0);
__decorate([
    cerialize.autoserializeAs('pan-entry'),
    __metadata("design:type", String)
], Filters.prototype, "panEntry", void 0);
__decorate([
    cerialize.autoserializeAs('cat-level'),
    __metadata("design:type", String)
], Filters.prototype, "catLevel", void 0);
__decorate([
    cerialize.autoserializeAs('acq-id'),
    __metadata("design:type", String)
], Filters.prototype, "acqId", void 0);
var DictionaryEntry = /** @class */ (function () {
    function DictionaryEntry() {
    }
    return DictionaryEntry;
}());
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], DictionaryEntry.prototype, "code", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], DictionaryEntry.prototype, "description", void 0);
var Dictionary = /** @class */ (function () {
    function Dictionary() {
    }
    Dictionary.OnDeserialized = function (instance, _) {
        if (!instance.value) {
            instance.parsed = {};
            return;
        }
        instance.parsed = instance.value.reduce(function (map, entry) {
            map[entry.code] = entry.description;
            return map;
        }, {});
    };
    return Dictionary;
}());
__decorate([
    cerialize.autoserializeAs(DictionaryEntry),
    __metadata("design:type", Array)
], Dictionary.prototype, "value", void 0);
var TccLimitAmount = /** @class */ (function () {
    function TccLimitAmount(initial) {
        Object.assign(this, initial);
    }
    return TccLimitAmount;
}());
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], TccLimitAmount.prototype, "tcc", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Number)
], TccLimitAmount.prototype, "amount", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], TccLimitAmount.prototype, "overridden", void 0);
var TccLimit = /** @class */ (function () {
    function TccLimit(initial) {
        Object.assign(this, initial);
    }
    TccLimit.OnSerialized = function (_, json) {
        json['brandProductCode'] = json['brandProductCode'].split('-')[0];
    };
    TccLimit.OnDeSerialized = function (_, json) {
        json.tccLimitAmount.forEach(function (item) {
            json.tccLimitAmount['tcc' + item.tcc] = item.amount;
            json.tccLimitAmount['overridden' + item.tcc] = item.overridden;
        });
    };
    return TccLimit;
}());
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Number)
], TccLimit.prototype, "tccLimitId", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Number)
], TccLimit.prototype, "memberId", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "startingAccntRange", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "endingAccntRange", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "brandProductCode", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Boolean)
], TccLimit.prototype, "vipIndicator", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "cardType", void 0);
__decorate([
    cerialize.autoserializeAs(TccLimitAmount),
    __metadata("design:type", Array)
], TccLimit.prototype, "tccLimitAmount", void 0);
__decorate([
    cerialize.autoserializeAs(Date),
    __metadata("design:type", Date)
], TccLimit.prototype, "lastUpdateDttm", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "lastUpdateUsr", void 0);
var DayDTO = /** @class */ (function () {
    function DayDTO() {
    }
    return DayDTO;
}());
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Number)
], DayDTO.prototype, "day", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Number)
], DayDTO.prototype, "amount", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Number)
], DayDTO.prototype, "count", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], DayDTO.prototype, "amountOverridden", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], DayDTO.prototype, "countOverridden", void 0);
var AccumLimits = /** @class */ (function () {
    function AccumLimits() {
    }
    return AccumLimits;
}());
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "memberId", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "startingAccntRange", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "endingAccntRange", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "brandProductCode", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "cardType", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Array)
], AccumLimits.prototype, "errors", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Boolean)
], AccumLimits.prototype, "changed", void 0);
__decorate([
    cerialize.autoserialize,
    __metadata("design:type", Array)
], AccumLimits.prototype, "accumLimitDay", void 0);
var FILTER_KEYS = Object.freeze({
    filters: 'filters',
    accumLimitsFilter: 'accumLimitsFilter',
    rangeBlockFilter: 'rangeBlockFilter',
    tccLimitsFilter: 'tccLimitsFilter',
    vipLimitsFilter: 'vipLimitsFilter',
});
var FILTER_CRITERIA = Object.freeze({
    startRange: 'start-range',
    endRange: 'end-range',
    memberId: 'member-id',
    posCtry: 'pos-ctry',
    mcc: 'mcc',
    createdOn: 'created-on',
    txType: 'tx-type',
    panEntry: 'pan-entry',
    catLevel: 'cat-level',
    acqId: 'acq-id',
    vipInd: 'vip-ind',
    cardType: 'card-type',
    brandProductCode: 'brand-product-code',
});
var SESSION_STORAGE = Object.freeze({
    cid: 'customerId',
    userId: 'userId'
});

exports.ErrorCodes = ErrorCodes;
exports.BlockCriteria = BlockCriteria;
exports.BlockType = BlockType;
exports.StandInBlock = StandInBlock;
exports.Filters = Filters;
exports.DictionaryEntry = DictionaryEntry;
exports.Dictionary = Dictionary;
exports.TccLimit = TccLimit;
exports.TccLimitAmount = TccLimitAmount;
exports.DayDTO = DayDTO;
exports.AccumLimits = AccumLimits;
exports.FILTER_KEYS = FILTER_KEYS;
exports.FILTER_CRITERIA = FILTER_CRITERIA;
exports.SESSION_STORAGE = SESSION_STORAGE;

Object.defineProperty(exports, '__esModule', { value: true });

})));
