// @mc-fraud-center/stand-in-data-model/esm5/mc-fraud-center-stand-in-data-model.js
import { __decorate, __metadata, __extends } from 'tslib';
import { autoserialize, deserialize, SerializableEnumeration, autoserializeAs, deserializeAs, inheritSerialization } from 'cerialize';
import * as momentNS from 'moment';

var ErrorCodes = Object.freeze({
    ACCT_RANGE_NOT_ALLOWED: 'acct.range.not.allowed',
    BILLING_LIMIT_EXCEED: 'billing.limit.exceed',
    BLOCK_ACCT_RANGE_TOO_WIDE: 'block.acct.range.too.wide',
    BLOCK_DUPLICATES_ICA_BLOCK: 'block.duplicates.ica.block',
    BLOCK_DUPLICATES_RANGE_BLOCK: 'block.duplicates.range.block',
    BLOCK_NOT_FOR_CUSTOMER: 'block.not.for.cust',
    BLOCK_SHADOWS_ICA_BLOCK: 'block.shadows.ica.block',
    BLOCK_SHADOWS_RANGE_BLOCK: 'block.shadows.range.block',
    ICA_NOT_ALLOWED: 'ica.not.allowed',
    INVALID_CUSTOMER: 'invalid.customer',
    INVALID_DATA_NOT_IN_DOMAIN: 'invalid.data.not.in.domain',
    INVALID_DATA_ORDER: 'invalid.data.order',
    INVALID_DATA_SIZE: 'invalid.data.size',
    INVALID_DATA_TYPE: 'invalid.data.type',
    INVALID_DATA_VALUE_FUTURE: 'invalid.date.value.future',
    INVALID_DATA_VALUE_PAST: 'invalid.date.value.past',
    INVALID_DECIMAL_VALUE_MAX: 'invalid.decimal.value.max',
    INVALID_DECIMAL_VALUE_MIN: 'invalid.decimal.value.min',
    INVALID_MAP_DUPLICATE_KEY: 'invalid.map.duplicate.key',
    INVALID_NOT_NULL_VALUE: 'invalid.not.null.value',
    INVALID_NUMERIC_VALUE_CURRENCY: 'invalid.numeric.value.currency',
    INVALID_NUMERIC_VALUE_DIGITS: 'invalid.numeric.value.digits',
    INVALID_NUMERIC_VALUE_MAX: 'invalid.numeric.value.max',
    INVALID_NUMERIC_VALUE_MIN: 'invalid.numeric.value.min',
    INVALID_NUMERIC_VALUE_RANGE: 'invalid.numeric.value.range',
    INVALID_TEXT_VALUE_BLANK: 'invalid.text.value.blank',
    INVALID_TEXT_VALUE_EMPTY: 'invalid.text.value.empty',
    INVALID_TEXT_VALUE_LENGTH: 'invalid.text.value.length',
    INVALID_TEXT_VALUE_PATTERN: 'invalid.text.value.pattern',
    REQUEST_INVALID_ILLEGAL_ARGUMENT: 'request.invalid.illegal.argument',
    REQUEST_INVALID_MESSAGE_NOT_READABLE: 'request.invalid.message.not.readable',
    REQUEST_INVALID_UNSATISFIED_REQUEST_PARAMETER: 'request.invalid.unsatisfied.request.parameter',
    TCC_LIMIT_ACCT_RANGE_NOT_FOUND: 'tcc.limit.acct.range.not.found',
    TCC_LIMIT_MINIMUM_THRESHOLD_VIOLATION: 'tcc.limit.min.threshold.violation',
    ACCUM_LIMIT_ACCT_RANGE_NOT_FOUND: 'accum.limit.acct.range.not.found',
    ACCUM_LIMIT_MIN_AMOUNT_THRESHOLD_VIOLATION: 'accum.limit.min.amount.threshold.violation',
    ACCUM_LIMIT_MAX_AMOUNT_THRESHOLD_VIOLATION: 'accum.limit.max.amount.threshold.violation',
    ACCUM_LIMIT_MIN_COUNT_THRESHOLD_VIOLATION: 'accum.limit.min.count.threshold.violation',
    ACCUM_LIMIT_MAX_COUNT_THRESHOLD_VIOLATION: 'accum.limit.max.count.threshold.violation',
    REQUEST_UNAUTHORIZED: 'request.unauthorized',
    ACCESS_DENIED_CSRF: 'access.denied.csrf',
    ACCESS_DENIED_PERMISSION: 'access.denied.permission',
    RESOURCE_NOT_FOUND: 'resource.not.found',
    REQUEST_INVALID_METHOD: 'request.invalid.method',
    REQUEST_INVALID_MEDIA_TYPE: 'request.invalid.media.type',
    REQUEST_CONFLICT: 'request.conflict',
    RESOURCE_GONE: 'resource.gone',
    REQUEST_UNSUPPORTED_MEDIA_TYPE: 'request.unsupported.media.type',
    SYSTEM: 'system',
    SERVICE_UNAVAILABLE: 'service.unavailable'
});
var BlockCriteria = /** @class */ (function () {
    function BlockCriteria(initial) {
        Object.assign(this, initial);
    }
    return BlockCriteria;
}());
__decorate([
    deserialize,
    __metadata("design:type", Number)
], BlockCriteria.prototype, "criteriaId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], BlockCriteria.prototype, "elementId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], BlockCriteria.prototype, "value", void 0);
var BlockType = {
    RANGE: 0,
    ICA: 1,
};
BlockType[BlockType.RANGE] = "RANGE";
BlockType[BlockType.ICA] = "ICA";
SerializableEnumeration(BlockType);
var StandInBlock = /** @class */ (function () {
    function StandInBlock(initial) {
        this.blockCriteria = [];
        Object.assign(this, initial);
    }
    StandInBlock.prototype.getBlockCriteria = function (id) {
        var found = this.blockCriteria && this.blockCriteria.find(function (item) { return item.elementId === id; });
        if (found) {
            return found;
        }
        var blockCriteria = new BlockCriteria();
        blockCriteria.elementId = id;
        blockCriteria.value = null;
        return blockCriteria;
    };
    return StandInBlock;
}());
__decorate([
    deserialize,
    __metadata("design:type", Number)
], StandInBlock.prototype, "blockId", void 0);
__decorate([
    deserializeAs(BlockType),
    __metadata("design:type", Number)
], StandInBlock.prototype, "blockType", void 0);
__decorate([
    deserialize,
    __metadata("design:type", Number)
], StandInBlock.prototype, "memberId", void 0);
__decorate([
    autoserializeAs(BlockCriteria),
    __metadata("design:type", Array)
], StandInBlock.prototype, "blockCriteria", void 0);
__decorate([
    deserializeAs(Date),
    __metadata("design:type", Date)
], StandInBlock.prototype, "lastUpdateDttm", void 0);
__decorate([
    deserialize,
    __metadata("design:type", String)
], StandInBlock.prototype, "lastUpdateUsr", void 0);
__decorate([
    deserializeAs(Date),
    __metadata("design:type", Date)
], StandInBlock.prototype, "crteDttm", void 0);
__decorate([
    deserialize,
    __metadata("design:type", String)
], StandInBlock.prototype, "crteUsr", void 0);
var RangeBlock = /** @class */ (function (_super) {
    __extends(RangeBlock, _super);
    function RangeBlock(initial) {
        var _this = _super.call(this, initial) || this;
        Object.assign(_this, initial);
        _this.blockType = BlockType.RANGE;
        return _this;
    }
    return RangeBlock;
}(StandInBlock));
__decorate([
    deserialize,
    __metadata("design:type", String)
], RangeBlock.prototype, "startingBlockRange", void 0);
__decorate([
    deserialize,
    __metadata("design:type", String)
], RangeBlock.prototype, "endingBlockRange", void 0);
RangeBlock = __decorate([
    inheritSerialization(StandInBlock),
    __metadata("design:paramtypes", [RangeBlock])
], RangeBlock);
var IcaBlock = /** @class */ (function (_super) {
    __extends(IcaBlock, _super);
    function IcaBlock(initial) {
        var _this = _super.call(this, initial) || this;
        Object.assign(_this, initial);
        _this.blockType = BlockType.ICA;
        return _this;
    }
    return IcaBlock;
}(StandInBlock));
IcaBlock = __decorate([
    inheritSerialization(StandInBlock),
    __metadata("design:paramtypes", [IcaBlock])
], IcaBlock);
var moment = momentNS;
var Filters = /** @class */ (function () {
    function Filters(initial) {
        Object.assign(this, initial);
    }
    Object.defineProperty(Filters.prototype, 'start-range', {
        get: function () { return this.startRange; },
        set: function (v) { this.startRange = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'end-range', {
        get: function () { return this.endRange; },
        set: function (v) { this.endRange = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'block-type', {
        get: function () { return this.blockType; },
        set: function (v) { this.blockType = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'member-id', {
        get: function () { return this.memberId; },
        set: function (v) { this.memberId = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'pos-ctry', {
        get: function () { return this.posCtry; },
        set: function (v) { this.posCtry = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'create-dttm-start', {
        get: function () { return this.createDttmStart; },
        set: function (v) { this.createDttmStart = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'create-dttm-end', {
        get: function () {
            if (this['create-dttm-start']) {
                this.createDttmEnd = moment(this['create-dttm-start']).add(1, 'days');
            }
            else {
                this.createDttmEnd = this['create-dttm-start'];
            }
            return this.createDttmEnd;
        },
        set: function (_) {
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'tx-type', {
        get: function () { return this.txType; },
        set: function (v) { this.txType = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'pan-entry', {
        get: function () { return this.panEntry; },
        set: function (v) { this.panEntry = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'cat-level', {
        get: function () { return this.catLevel; },
        set: function (v) { this.catLevel = v; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Filters.prototype, 'acq-id', {
        get: function () { return this.acqId; },
        set: function (v) { this.acqId = v; },
        enumerable: true,
        configurable: true
    });
    return Filters;
}());
__decorate([
    autoserialize,
    __metadata("design:type", String)
], Filters.prototype, "mcc", void 0);
__decorate([
    autoserializeAs('start-range'),
    __metadata("design:type", String)
], Filters.prototype, "startRange", void 0);
__decorate([
    autoserializeAs('end-range'),
    __metadata("design:type", String)
], Filters.prototype, "endRange", void 0);
__decorate([
    autoserializeAs('block-type'),
    __metadata("design:type", String)
], Filters.prototype, "blockType", void 0);
__decorate([
    autoserializeAs('member-id'),
    __metadata("design:type", String)
], Filters.prototype, "memberId", void 0);
__decorate([
    autoserializeAs('pos-ctry'),
    __metadata("design:type", String)
], Filters.prototype, "posCtry", void 0);
__decorate([
    autoserializeAs(Date, 'create-dttm-start'),
    __metadata("design:type", Date)
], Filters.prototype, "createDttmStart", void 0);
__decorate([
    autoserializeAs(Date, 'create-dttm-end'),
    __metadata("design:type", Object)
], Filters.prototype, "createDttmEnd", void 0);
__decorate([
    autoserializeAs('tx-type'),
    __metadata("design:type", String)
], Filters.prototype, "txType", void 0);
__decorate([
    autoserializeAs('pan-entry'),
    __metadata("design:type", String)
], Filters.prototype, "panEntry", void 0);
__decorate([
    autoserializeAs('cat-level'),
    __metadata("design:type", String)
], Filters.prototype, "catLevel", void 0);
__decorate([
    autoserializeAs('acq-id'),
    __metadata("design:type", String)
], Filters.prototype, "acqId", void 0);
var DictionaryEntry = /** @class */ (function () {
    function DictionaryEntry() {
    }
    return DictionaryEntry;
}());
__decorate([
    autoserialize,
    __metadata("design:type", String)
], DictionaryEntry.prototype, "code", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], DictionaryEntry.prototype, "description", void 0);
var Dictionary = /** @class */ (function () {
    function Dictionary() {
    }
    Dictionary.OnDeserialized = function (instance, _) {
        if (!instance.value) {
            instance.parsed = {};
            return;
        }
        instance.parsed = instance.value.reduce(function (map, entry) {
            map[entry.code] = entry.description;
            return map;
        }, {});
    };
    return Dictionary;
}());
__decorate([
    autoserializeAs(DictionaryEntry),
    __metadata("design:type", Array)
], Dictionary.prototype, "value", void 0);
var TccLimitAmount = /** @class */ (function () {
    function TccLimitAmount(initial) {
        Object.assign(this, initial);
    }
    return TccLimitAmount;
}());
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimitAmount.prototype, "tcc", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], TccLimitAmount.prototype, "amount", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimitAmount.prototype, "overridden", void 0);
var TccLimit = /** @class */ (function () {
    function TccLimit(initial) {
        Object.assign(this, initial);
    }
    TccLimit.OnSerialized = function (_, json) {
        json['brandProductCode'] = json['brandProductCode'].split('-')[0];
    };
    TccLimit.OnDeSerialized = function (_, json) {
        json.tccLimitAmount.forEach(function (item) {
            json.tccLimitAmount['tcc' + item.tcc] = item.amount;
            json.tccLimitAmount['overridden' + item.tcc] = item.overridden;
        });
    };
    return TccLimit;
}());
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], TccLimit.prototype, "tccLimitId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], TccLimit.prototype, "memberId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "startingAccntRange", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "endingAccntRange", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "brandProductCode", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Boolean)
], TccLimit.prototype, "vipIndicator", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "cardType", void 0);
__decorate([
    autoserializeAs(TccLimitAmount),
    __metadata("design:type", Array)
], TccLimit.prototype, "tccLimitAmount", void 0);
__decorate([
    autoserializeAs(Date),
    __metadata("design:type", Date)
], TccLimit.prototype, "lastUpdateDttm", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "lastUpdateUsr", void 0);
var DayDTO = /** @class */ (function () {
    function DayDTO() {
    }
    return DayDTO;
}());
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], DayDTO.prototype, "day", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], DayDTO.prototype, "amount", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], DayDTO.prototype, "count", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], DayDTO.prototype, "amountOverridden", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], DayDTO.prototype, "countOverridden", void 0);
var AccumLimits = /** @class */ (function () {
    function AccumLimits() {
    }
    return AccumLimits;
}());
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "memberId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "startingAccntRange", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "endingAccntRange", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "brandProductCode", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "cardType", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Array)
], AccumLimits.prototype, "errors", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Boolean)
], AccumLimits.prototype, "changed", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Array)
], AccumLimits.prototype, "accumLimitDay", void 0);
var FILTER_KEYS = Object.freeze({
    filters: 'filters',
    accumLimitsFilter: 'accumLimitsFilter',
    rangeBlockFilter: 'rangeBlockFilter',
    tccLimitsFilter: 'tccLimitsFilter',
    vipLimitsFilter: 'vipLimitsFilter',
});
var FILTER_CRITERIA = Object.freeze({
    startRange: 'start-range',
    endRange: 'end-range',
    memberId: 'member-id',
    posCtry: 'pos-ctry',
    mcc: 'mcc',
    createdOn: 'created-on',
    txType: 'tx-type',
    panEntry: 'pan-entry',
    catLevel: 'cat-level',
    acqId: 'acq-id',
    vipInd: 'vip-ind',
    cardType: 'card-type',
    brandProductCode: 'brand-product-code',
});
var SESSION_STORAGE = Object.freeze({
    cid: 'customerId',
    userId: 'userId'
});

export { ErrorCodes, BlockCriteria, BlockType, RangeBlock, IcaBlock, StandInBlock, Filters, DictionaryEntry, Dictionary, TccLimit, TccLimitAmount, DayDTO, AccumLimits, FILTER_KEYS, FILTER_CRITERIA, SESSION_STORAGE };
