// @mc-fraud-center/stand-in-data-model/esm2015/mc-fraud-center-stand-in-data-model.js
import { __decorate, __metadata } from 'tslib';
import { autoserialize, deserialize, SerializableEnumeration, autoserializeAs, deserializeAs, inheritSerialization } from 'cerialize';
import * as momentNS from 'moment';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Represents all the available error codes response of an API request in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/viewpage.action?pageId=222361795
 */
const ErrorCodes = Object.freeze({
    // HTTP STATUS 400
    ACCT_RANGE_NOT_ALLOWED: 'acct.range.not.allowed',
    BILLING_LIMIT_EXCEED: 'billing.limit.exceed',
    BLOCK_ACCT_RANGE_TOO_WIDE: 'block.acct.range.too.wide',
    BLOCK_DUPLICATES_ICA_BLOCK: 'block.duplicates.ica.block',
    BLOCK_DUPLICATES_RANGE_BLOCK: 'block.duplicates.range.block',
    BLOCK_NOT_FOR_CUSTOMER: 'block.not.for.cust',
    BLOCK_SHADOWS_ICA_BLOCK: 'block.shadows.ica.block',
    BLOCK_SHADOWS_RANGE_BLOCK: 'block.shadows.range.block',
    ICA_NOT_ALLOWED: 'ica.not.allowed',
    INVALID_CUSTOMER: 'invalid.customer',
    INVALID_DATA_NOT_IN_DOMAIN: 'invalid.data.not.in.domain',
    INVALID_DATA_ORDER: 'invalid.data.order',
    INVALID_DATA_SIZE: 'invalid.data.size',
    INVALID_DATA_TYPE: 'invalid.data.type',
    INVALID_DATA_VALUE_FUTURE: 'invalid.date.value.future',
    INVALID_DATA_VALUE_PAST: 'invalid.date.value.past',
    INVALID_DECIMAL_VALUE_MAX: 'invalid.decimal.value.max',
    INVALID_DECIMAL_VALUE_MIN: 'invalid.decimal.value.min',
    INVALID_MAP_DUPLICATE_KEY: 'invalid.map.duplicate.key',
    INVALID_NOT_NULL_VALUE: 'invalid.not.null.value',
    INVALID_NUMERIC_VALUE_CURRENCY: 'invalid.numeric.value.currency',
    INVALID_NUMERIC_VALUE_DIGITS: 'invalid.numeric.value.digits',
    INVALID_NUMERIC_VALUE_MAX: 'invalid.numeric.value.max',
    INVALID_NUMERIC_VALUE_MIN: 'invalid.numeric.value.min',
    INVALID_NUMERIC_VALUE_RANGE: 'invalid.numeric.value.range',
    INVALID_TEXT_VALUE_BLANK: 'invalid.text.value.blank',
    INVALID_TEXT_VALUE_EMPTY: 'invalid.text.value.empty',
    INVALID_TEXT_VALUE_LENGTH: 'invalid.text.value.length',
    INVALID_TEXT_VALUE_PATTERN: 'invalid.text.value.pattern',
    REQUEST_INVALID_ILLEGAL_ARGUMENT: 'request.invalid.illegal.argument',
    REQUEST_INVALID_MESSAGE_NOT_READABLE: 'request.invalid.message.not.readable',
    REQUEST_INVALID_UNSATISFIED_REQUEST_PARAMETER: 'request.invalid.unsatisfied.request.parameter',
    TCC_LIMIT_ACCT_RANGE_NOT_FOUND: 'tcc.limit.acct.range.not.found',
    TCC_LIMIT_MINIMUM_THRESHOLD_VIOLATION: 'tcc.limit.min.threshold.violation',
    // ACCUM_LIMITS
    ACCUM_LIMIT_ACCT_RANGE_NOT_FOUND: 'accum.limit.acct.range.not.found',
    ACCUM_LIMIT_MIN_AMOUNT_THRESHOLD_VIOLATION: 'accum.limit.min.amount.threshold.violation',
    ACCUM_LIMIT_MAX_AMOUNT_THRESHOLD_VIOLATION: 'accum.limit.max.amount.threshold.violation',
    ACCUM_LIMIT_MIN_COUNT_THRESHOLD_VIOLATION: 'accum.limit.min.count.threshold.violation',
    ACCUM_LIMIT_MAX_COUNT_THRESHOLD_VIOLATION: 'accum.limit.max.count.threshold.violation',
    // HTTP STATUS 401
    REQUEST_UNAUTHORIZED: 'request.unauthorized',
    // HTTP STATUS 403
    ACCESS_DENIED_CSRF: 'access.denied.csrf',
    ACCESS_DENIED_PERMISSION: 'access.denied.permission',
    // HTTP STATUS 404
    RESOURCE_NOT_FOUND: 'resource.not.found',
    // HTTP STATUS 405
    REQUEST_INVALID_METHOD: 'request.invalid.method',
    // HTTP STATUS 406
    REQUEST_INVALID_MEDIA_TYPE: 'request.invalid.media.type',
    // HTTP STATUS 409
    REQUEST_CONFLICT: 'request.conflict',
    // HTTP STATUS 410
    RESOURCE_GONE: 'resource.gone',
    // HTTP STATUS 415
    REQUEST_UNSUPPORTED_MEDIA_TYPE: 'request.unsupported.media.type',
    // HTTP STATUS 500
    SYSTEM: 'system',
    // HTTP STATUS 503
    SERVICE_UNAVAILABLE: 'service.unavailable'
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Represents a view of a Range Block Criteria resource in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/tinyurl.action?urlIdentifier=lIPyD
 */
class BlockCriteria {
    /**
     * @param {?=} initial
     */
    constructor(initial) {
        Object.assign(this, initial);
    }
}
__decorate([
    deserialize,
    __metadata("design:type", Number)
], BlockCriteria.prototype, "criteriaId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], BlockCriteria.prototype, "elementId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], BlockCriteria.prototype, "value", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {number} */
const BlockType = {
    RANGE: 0,
    ICA: 1,
};
BlockType[BlockType.RANGE] = "RANGE";
BlockType[BlockType.ICA] = "ICA";
SerializableEnumeration(BlockType);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Represents a generic view of a Range Block or ICA Block in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/tinyurl.action?urlIdentifier=lIPyD
 */
class StandInBlock {
    /**
     * @param {?=} initial
     */
    constructor(initial) {
        /**
         * An array of BlockCriteria resources associated to this block.
         */
        this.blockCriteria = [];
        Object.assign(this, initial);
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getBlockCriteria(id) {
        const /** @type {?} */ found = this.blockCriteria && this.blockCriteria.find((item) => item.elementId === id);
        if (found) {
            return found;
        }
        const /** @type {?} */ blockCriteria = new BlockCriteria();
        blockCriteria.elementId = id;
        blockCriteria.value = null;
        return blockCriteria;
    }
}
__decorate([
    deserialize,
    __metadata("design:type", Number)
], StandInBlock.prototype, "blockId", void 0);
__decorate([
    deserializeAs(BlockType),
    __metadata("design:type", Number)
], StandInBlock.prototype, "blockType", void 0);
__decorate([
    deserialize,
    __metadata("design:type", Number)
], StandInBlock.prototype, "memberId", void 0);
__decorate([
    autoserializeAs(BlockCriteria),
    __metadata("design:type", Array)
], StandInBlock.prototype, "blockCriteria", void 0);
__decorate([
    deserializeAs(Date),
    __metadata("design:type", Date)
], StandInBlock.prototype, "lastUpdateDttm", void 0);
__decorate([
    deserialize,
    __metadata("design:type", String)
], StandInBlock.prototype, "lastUpdateUsr", void 0);
__decorate([
    deserializeAs(Date),
    __metadata("design:type", Date)
], StandInBlock.prototype, "crteDttm", void 0);
__decorate([
    deserialize,
    __metadata("design:type", String)
], StandInBlock.prototype, "crteUsr", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Represents a generic view of a Range Block in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/tinyurl.action?urlIdentifier=lIPyD
 */
let RangeBlock = /**
 * Represents a generic view of a Range Block in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/tinyurl.action?urlIdentifier=lIPyD
 */
class RangeBlock extends StandInBlock {
    /**
     * @param {?=} initial
     */
    constructor(initial) {
        super(initial);
        Object.assign(this, initial);
        this.blockType = BlockType.RANGE;
    }
};
__decorate([
    deserialize,
    __metadata("design:type", String)
], RangeBlock.prototype, "startingBlockRange", void 0);
__decorate([
    deserialize,
    __metadata("design:type", String)
], RangeBlock.prototype, "endingBlockRange", void 0);
/**
 * Represents a generic view of a Range Block in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/tinyurl.action?urlIdentifier=lIPyD
 */
RangeBlock = __decorate([
    inheritSerialization(StandInBlock),
    __metadata("design:paramtypes", [RangeBlock])
], RangeBlock);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Represents a generic view of a ICA Block in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/tinyurl.action?urlIdentifier=lIPyD
 */
let IcaBlock = /**
 * Represents a generic view of a ICA Block in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/tinyurl.action?urlIdentifier=lIPyD
 */
class IcaBlock extends StandInBlock {
    /**
     * @param {?=} initial
     */
    constructor(initial) {
        super(initial);
        Object.assign(this, initial);
        this.blockType = BlockType.ICA;
    }
};
/**
 * Represents a generic view of a ICA Block in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/tinyurl.action?urlIdentifier=lIPyD
 */
IcaBlock = __decorate([
    inheritSerialization(StandInBlock),
    __metadata("design:paramtypes", [IcaBlock])
], IcaBlock);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const moment = momentNS;
class Filters {
    /**
     * @param {?=} initial
     */
    constructor(initial) {
        Object.assign(this, initial);
    }
    /**
     * @return {?}
     */
    get 'start-range'() { return this.startRange; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'start-range'(v) { this.startRange = v; }
    /**
     * @return {?}
     */
    get 'end-range'() { return this.endRange; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'end-range'(v) { this.endRange = v; }
    /**
     * @return {?}
     */
    get 'block-type'() { return this.blockType; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'block-type'(v) { this.blockType = v; }
    /**
     * @return {?}
     */
    get 'member-id'() { return this.memberId; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'member-id'(v) { this.memberId = v; }
    /**
     * @return {?}
     */
    get 'pos-ctry'() { return this.posCtry; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'pos-ctry'(v) { this.posCtry = v; }
    /**
     * @return {?}
     */
    get 'create-dttm-start'() { return this.createDttmStart; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'create-dttm-start'(v) { this.createDttmStart = v; }
    /**
     * @return {?}
     */
    get 'create-dttm-end'() {
        if (this['create-dttm-start']) {
            this.createDttmEnd = moment(this['create-dttm-start']).add(1, 'days');
        }
        else {
            this.createDttmEnd = this['create-dttm-start'];
        }
        return this.createDttmEnd;
    }
    /**
     * @param {?} _
     * @return {?}
     */
    set 'create-dttm-end'(_) {
        /* noop */
    }
    /**
     * @return {?}
     */
    get 'tx-type'() { return this.txType; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'tx-type'(v) { this.txType = v; }
    /**
     * @return {?}
     */
    get 'pan-entry'() { return this.panEntry; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'pan-entry'(v) { this.panEntry = v; }
    /**
     * @return {?}
     */
    get 'cat-level'() { return this.catLevel; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'cat-level'(v) { this.catLevel = v; }
    /**
     * @return {?}
     */
    get 'acq-id'() { return this.acqId; }
    /**
     * @param {?} v
     * @return {?}
     */
    set 'acq-id'(v) { this.acqId = v; }
}
__decorate([
    autoserialize,
    __metadata("design:type", String)
], Filters.prototype, "mcc", void 0);
__decorate([
    autoserializeAs('start-range'),
    __metadata("design:type", String)
], Filters.prototype, "startRange", void 0);
__decorate([
    autoserializeAs('end-range'),
    __metadata("design:type", String)
], Filters.prototype, "endRange", void 0);
__decorate([
    autoserializeAs('block-type'),
    __metadata("design:type", String)
], Filters.prototype, "blockType", void 0);
__decorate([
    autoserializeAs('member-id'),
    __metadata("design:type", String)
], Filters.prototype, "memberId", void 0);
__decorate([
    autoserializeAs('pos-ctry'),
    __metadata("design:type", String)
], Filters.prototype, "posCtry", void 0);
__decorate([
    autoserializeAs(Date, 'create-dttm-start'),
    __metadata("design:type", Date)
], Filters.prototype, "createDttmStart", void 0);
__decorate([
    autoserializeAs(Date, 'create-dttm-end'),
    __metadata("design:type", Object)
], Filters.prototype, "createDttmEnd", void 0);
__decorate([
    autoserializeAs('tx-type'),
    __metadata("design:type", String)
], Filters.prototype, "txType", void 0);
__decorate([
    autoserializeAs('pan-entry'),
    __metadata("design:type", String)
], Filters.prototype, "panEntry", void 0);
__decorate([
    autoserializeAs('cat-level'),
    __metadata("design:type", String)
], Filters.prototype, "catLevel", void 0);
__decorate([
    autoserializeAs('acq-id'),
    __metadata("design:type", String)
], Filters.prototype, "acqId", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class DictionaryEntry {
}
__decorate([
    autoserialize,
    __metadata("design:type", String)
], DictionaryEntry.prototype, "code", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], DictionaryEntry.prototype, "description", void 0);
class Dictionary {
    /**
     * @param {?} instance
     * @param {?} _
     * @return {?}
     */
    static OnDeserialized(instance, _) {
        if (!instance.value) {
            instance.parsed = {};
            return;
        }
        instance.parsed = instance.value.reduce((map, entry) => {
            map[entry.code] = entry.description;
            return map;
        }, {});
    }
}
__decorate([
    autoserializeAs(DictionaryEntry),
    __metadata("design:type", Array)
], Dictionary.prototype, "value", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 *  represents the limit amount for each of the 11 TCC's and whether the amounts are different than default limits.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/viewpage.action?pageId=221124414
 */
class TccLimitAmount {
    /**
     * @param {?=} initial
     */
    constructor(initial) {
        Object.assign(this, initial);
    }
}
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimitAmount.prototype, "tcc", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], TccLimitAmount.prototype, "amount", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimitAmount.prototype, "overridden", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Represents a generic view of a Range Block or ICA Block in the system.
 * @see https://dtl.fusion.proxy.mastercard.int/confluence/pages/tinyurl.action?urlIdentifier=lIPyD
 */
class TccLimit {
    /**
     * @param {?=} initial
     */
    constructor(initial) {
        Object.assign(this, initial);
    }
    /**
     * @param {?} _
     * @param {?} json
     * @return {?}
     */
    static OnSerialized(_, json) {
        json['brandProductCode'] = json['brandProductCode'].split('-')[0];
    }
    /**
     * @param {?} _
     * @param {?} json
     * @return {?}
     */
    static OnDeSerialized(_, json) {
        json.tccLimitAmount.forEach((item) => {
            json.tccLimitAmount['tcc' + item.tcc] = item.amount;
            json.tccLimitAmount['overridden' + item.tcc] = item.overridden;
        });
    }
}
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], TccLimit.prototype, "tccLimitId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], TccLimit.prototype, "memberId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "startingAccntRange", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "endingAccntRange", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "brandProductCode", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Boolean)
], TccLimit.prototype, "vipIndicator", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "cardType", void 0);
__decorate([
    autoserializeAs(TccLimitAmount),
    __metadata("design:type", Array)
], TccLimit.prototype, "tccLimitAmount", void 0);
__decorate([
    autoserializeAs(Date),
    __metadata("design:type", Date)
], TccLimit.prototype, "lastUpdateDttm", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], TccLimit.prototype, "lastUpdateUsr", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class DayDTO {
}
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], DayDTO.prototype, "day", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], DayDTO.prototype, "amount", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Number)
], DayDTO.prototype, "count", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], DayDTO.prototype, "amountOverridden", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], DayDTO.prototype, "countOverridden", void 0);
class AccumLimits {
}
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "memberId", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "startingAccntRange", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "endingAccntRange", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "brandProductCode", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", String)
], AccumLimits.prototype, "cardType", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Array)
], AccumLimits.prototype, "errors", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Boolean)
], AccumLimits.prototype, "changed", void 0);
__decorate([
    autoserialize,
    __metadata("design:type", Array)
], AccumLimits.prototype, "accumLimitDay", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const FILTER_KEYS = Object.freeze({
    filters: 'filters',
    accumLimitsFilter: 'accumLimitsFilter',
    rangeBlockFilter: 'rangeBlockFilter',
    tccLimitsFilter: 'tccLimitsFilter',
    vipLimitsFilter: 'vipLimitsFilter',
});
const FILTER_CRITERIA = Object.freeze({
    startRange: 'start-range',
    endRange: 'end-range',
    memberId: 'member-id',
    posCtry: 'pos-ctry',
    mcc: 'mcc',
    createdOn: 'created-on',
    txType: 'tx-type',
    panEntry: 'pan-entry',
    catLevel: 'cat-level',
    acqId: 'acq-id',
    vipInd: 'vip-ind',
    cardType: 'card-type',
    brandProductCode: 'brand-product-code',
});
const SESSION_STORAGE = Object.freeze({
    cid: 'customerId',
    userId: 'userId'
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/* API errors */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { ErrorCodes, BlockCriteria, BlockType, RangeBlock, IcaBlock, StandInBlock, Filters, DictionaryEntry, Dictionary, TccLimit, TccLimitAmount, DayDTO, AccumLimits, FILTER_KEYS, FILTER_CRITERIA, SESSION_STORAGE };
